### PR TITLE
fix(backtesting): correct regime fallback and probabilities

### LINKED ISSUE #247: [BUG] analyze_market_regimes fabricates uniform probabilities and misreports method after silent fit fallback
## 🐛 Bug Description

`backtesting_analyze_market_regimes` can silently run entirely on the rule-based threshold classifier while still reporting `method: "hmm"`/`"kmeans"` and returning a fabricated uniform `[0.333, 0.333, 0.333]` probability distribution for every day in the history, regardless of the regime actually assigned.

Root cause: `MarketRegimeDetector.fit_regimes` has several "not enough data" early-return branches (e.g. fewer than `max(50, n_regimes*20)` valid feature windows -- which the tool's own default 365-day lookback usually does not clear for `n_regimes=3`). Each of these sets `self.is_fitted = True` **without ever fitting `self.scaler`/`self.model`**. Downstream, `get_regime_probabilities` then hits `self.scaler.transform(...)`, raises `NotFittedError` on the never-fitted scaler, and its broad `except` fabricates `np.ones(n) / n`. Independently, `detect_current_regime` falls back to the rule-based threshold classifier through its own, separately-triggered exception handling -- so the regime *label* varies plausibly day-to-day while the *probability* stays fake-uniform, and `self.method` is never updated, so the tool keeps reporting the originally-requested method.

## 💰 Financial Disclaimer Acknowledgment

- [x] I understand this is educational software and not financial advice
- [x] I am not expecting investment recommendations or guaranteed returns
- [x] This bug report is about technical functionality, not financial performance

## 📋 Reproduction Steps

1. Call `backtesting_analyze_market_regimes` with `symbol="AAPL"`, `method="hmm"`, `n_regimes=3`, using the tool's default lookback (no `start_date`/`end_date`).
2. Inspect the returned `method` field and every `recent_regime_history[*].probabilities` entry.

## 🎯 Expected Behavior

`method` should report what was actually used (`"threshold"` when the fit fell back), and `probabilities` should either be the fitted model's real posterior or an honest one-hot vector at the assigned regime.

**Actual behavior**: `method` still reports `"hmm"`, and every `probabilities` entry is exactly `[0.3333333333333333, 0.3333333333333333, 0.3333333333333333]` regardless of which regime was assigned that day.

Separately: `method="hmm"` is also a misnomer -- it fits `sklearn.mixture.GaussianMixture` (static clustering, no transition-probability/temporal modeling), not a genuine Hidden Markov Model. This should be documented even where a real fit succeeds.

## 💻 Environment Information

- OS: Windows 11
- Python Version: 3.12.10
- MaverickMCP Version: main branch, HEAD `95c919f8ed22aa4dd0b58525900769c30a32ad3d` at time of report
- Installation Method: `uv sync --extra backtesting`

## 📊 Error Messages/Logs

No exception is raised. The tool returns a successful response with a misleading `method` field and fabricated confidence values.

## 🔧 Additional Context

Confirmed by direct source inspection of `maverick/backtesting/strategies/ml/regime_detector.py::MarketRegimeDetector` and by empirical reproduction (instrumented calls showing `NotFittedError` being swallowed). A fix, accurate documentation, and regression tests are proposed in PR (linked from this issue).

## 🏷️ Bug Classification

**Severity:**
- [x] Medium (feature partially working) -- silently fabricates analytical confidence

**Component:**
- [x] MCP server/tools (backtesting domain)

### PR DESCRIPTION
**Pull request by A1-NWS-Dev1**
## 📋 Pull Request Summary

**Brief description of changes:**
Fixes `backtesting_analyze_market_regimes`'s silent fit-state dishonesty: it could report `method: "hmm"`/`"kmeans"` and a fabricated uniform probability distribution while actually running entirely on the rule-based threshold classifier.

**Related issue(s):**
- Fixes #247

## 💰 Financial Disclaimer Acknowledgment

- [x] I understand this is educational software and not financial advice
- [x] Any financial analysis features include appropriate disclaimers
- [x] This PR maintains the educational/personal-use focus of the project

## 🔄 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update (improvements to documentation)
- [ ] 🔧 Refactor (code changes that neither fix bugs nor add features)
- [ ] ⚡ Performance improvement
- [ ] 🧹 Chore (dependencies, build tools, etc.)

## 🎯 Component Areas

**Primary areas affected:**
- [ ] Data fetching (Tiingo, Yahoo Finance, FRED)
- [ ] Technical analysis calculations
- [ ] Stock screening strategies
- [ ] Portfolio analysis and optimization
- [x] MCP server/tools implementation (backtesting domain)
- [ ] Database operations and models
- [ ] Caching (Redis/in-memory)
- [ ] Claude Desktop integration
- [ ] Development tools and setup
- [x] Documentation and examples

## 🔧 Implementation Details

**Technical approach:**
`MarketRegimeDetector.fit_regimes` has several "not enough data" early-return branches (e.g. fewer than `max(50, n_regimes*20)` valid feature windows -- which this tool's own default 365-day lookback usually doesn't clear for `n_regimes=3`). Each of these set `self.is_fitted = True` without ever fitting `self.scaler`/`self.model`. `get_regime_probabilities` then hit `self.scaler.transform(...)`, raised `NotFittedError` on the never-fitted scaler, and its broad `except` fabricated `np.ones(n) / n`. Independently, `detect_current_regime` fell back to the rule-based threshold classifier through its own, separately-triggered exception handling -- so the regime label varied plausibly while the probability stayed fake-uniform, and `self.method` was never updated.

**Key changes:**
- Every "can't fit a genuine statistical model" branch in `fit_regimes` (four early-return conditions plus two exception handlers) now converges on one explicit state already used for an explicit `method="threshold"` request: `self.method = "threshold"; self.is_fitted = True`.
- `get_regime_probabilities` now returns the fitted model's real `predict_proba` posterior when genuinely available, or an honest one-hot vector at the actually-assigned regime -- the fabricated-uniform return path is removed entirely.
- `analyze_market_regimes` now reports `detector.method` (what was actually used) instead of the original request.
- `method="hmm"` remains an accepted, backward-compatible input value; docstrings now state clearly it fits `sklearn.mixture.GaussianMixture` (no transition/temporal modeling), not a real Hidden Markov Model. No new dependency introduced.
- Found via CodeRabbit review on this PR: `_one_hot_via_threshold_fallback` indexed `probs[regime]` without checking it against `self.n_regimes`. `detect_regime_threshold` always returns one of exactly three hardcoded labels (0/1/2), so a detector configured with `n_regimes < 3` could raise `IndexError`. The label is now clamped into `[0, n_regimes)` before indexing.
- Also from CodeRabbit review: tightened a regression test that only checked `sum()==1`/`max()==1` on returned probabilities (which a non-one-hot vector could also satisfy) to compare the exact expected one-hot vector.

**Dependencies:**
- [x] No new dependencies added

## 🧪 Testing

**Testing performed:**
- [x] Unit tests added/updated
- [ ] Integration tests added/updated
- [x] Manual testing completed (live MCP tool calls: default-lookback AAPL confirming honest `"threshold"` fallback + one-hot output; ~3.5-year lookback confirming a genuine `"hmm"` fit with real non-uniform posteriors)
- [ ] Tested with Claude Desktop
- [ ] Tested with different data sources
- [ ] Performance testing completed

**Test scenarios covered:**
- [x] Happy path functionality (genuine model fit on well-separated synthetic regimes)
- [x] Error handling and edge cases (insufficient data, runtime failure after a genuine fit, `n_regimes < 3`)
- [ ] Data validation and sanitization
- [ ] API rate limiting compliance
- [ ] Database operations
- [ ] Cache behavior

**Manual testing:**
```bash
uv run pytest tests/backtesting/ tests/server/
uv run --extra dev ruff check .
uv run --extra dev ruff format --check .
uv run lint-imports
uv run python tools/check_docs_catalog.py
```
Result: 245 passed, 5 skipped (pre-existing, Research-extra-gated), 0 failed. `lint-imports`: 14/14 contracts kept. `ruff check`/`ruff format --check`: clean. Docs catalog check: passed.

`make typecheck` could not be verified in this environment: it requires `--extra research` to resolve `maverick.backtesting.strategies.parser`'s lazy BYOK-LLM import (a pre-existing condition unrelated to this PR -- `parser.py` is untouched by this diff), and this environment does not install the `[research]` extra. Not claiming this check as passed.

## 📊 Financial Analysis Impact

**Financial calculations:**
- [ ] No financial calculations affected
- [ ] New financial calculations added (validated for accuracy)
- [x] Existing calculations modified (thoroughly tested)
- [x] All calculations include appropriate disclaimers

**Data providers:**
- [x] No data provider changes

**Market data handling:**
- [ ] Historical data processing
- [ ] Real-time data integration
- [x] Technical indicator calculations (regime classification confidence reporting)
- [ ] Screening algorithm changes

## 🔒 Security Considerations

**Security checklist:**
- [x] No hardcoded secrets or credentials
- [x] Input validation implemented (regime index now bounds-checked against `n_regimes`)
- [x] Error handling doesn't leak sensitive information
- [x] API keys handled securely via environment variables (unaffected)
- [x] SQL injection prevention verified (not applicable -- no SQL in this path)
- [x] Rate limiting respected for external APIs (unaffected)

## 📚 Documentation

**Documentation updates:**
- [x] Code comments added/updated
- [ ] README.md updated (if needed)
- [x] API documentation updated (`docs/api/backtesting.md`)
- [ ] Examples/tutorials added
- [x] Financial disclaimers included where appropriate

**Breaking changes documentation:**
- [x] No breaking changes (return shape unchanged; `method`/`probabilities` values become honest rather than fabricated)

## ✅ Pre-submission Checklist

**Code quality:**
- [x] Code follows the project style guide
- [x] Self-review of code completed
- [x] Tests added/updated and passing
- [x] No linting errors (`make lint` -- verified via `ruff check`/`ruff format --check`/`lint-imports`)
- [ ] Type checking passes (`make typecheck`) -- not run in this environment; see Testing section
- [x] All tests pass (`make test`)

**Financial software standards:**
- [x] Financial disclaimers included where appropriate
- [x] No investment advice or guarantees provided
- [x] Educational purpose maintained
- [x] Data accuracy considerations documented
- [x] Risk warnings included for relevant features

**Community standards:**
- [x] PR title is descriptive and follows convention
- [x] Description clearly explains the changes
- [x] Related issues are linked
- [ ] Screenshots/examples included (if applicable) -- not applicable, backend logic fix with no UI
- [x] Ready for review

## 🤝 Review Guidance

**Areas needing special attention:**
- A real Hidden Markov Model implementation is intentionally out of scope for this fix (no new heavy dependency); `method="hmm"` stays a documented misnomer for `GaussianMixture`, kept for backward compatibility.
- `detect_regime_threshold`'s hardcoded 3-way (0/1/2) output regardless of `n_regimes` is pre-existing legacy behavior; this PR only guards the one new call site (`_one_hot_via_threshold_fallback`) against indexing out of bounds, it does not redesign the threshold classifier itself.

## 🚀 Deployment Notes

**Environment considerations:**
- [x] No environment changes required

**Rollback plan:**
- [x] Changes are fully backward compatible

## 🎓 Educational Impact

**Learning value:**
Demonstrates how two independent, differently-triggered silent fallbacks for the same underlying condition can drift apart (one producing plausible-looking output, the other fabricating confidence) unless the fit-state itself is made the single source of truth.

---

**Additional Notes:**

## Tests added

- `test_well_separated_regimes_produce_non_uniform_probabilities`
- `test_hmm_method_is_gaussian_mixture_not_hidden_markov_model`
- `test_probability_lookup_failure_falls_back_to_one_hot_not_uniform`
- `test_threshold_fallback_does_not_index_error_with_fewer_than_three_regimes` (added for the CodeRabbit-found bounds issue)
- `test_analyze_market_regimes_reports_actual_fallback_method`
- Updated `test_insufficient_data_falls_back_to_threshold` to assert the new honest `is_fitted=True, method="threshold"` state

### CODE REVIEW

**Comment by A1-NWS-Dev1** [comment]
Updated this PR for conformity with CONTRIBUTING.md and .github/pull_request_template.md: linked issue #247, restructured the description to the repository's actual PR template, pushed a focused follow-up commit addressing all three CodeRabbit findings (n_regimes bounds guard in the threshold fallback, a tightened one-hot test assertion, and line-length wrapping), and re-ran the full test/lint/format/import-linter/docs-catalog suite (results in the PR description).

**Comment by wshobson** [comment]
Rebasing onto main after #243 and adding one fixup commit: (1) maverick/ is under a 500-line-per-file structural rule (tests/structure/test_harness_rules.py), and the multi-line comment above method=detector.method pushed service_ml.py to 505 lines on the pre-rebase base, so it becomes a one-line trailing comment; (2) the two GaussianMixture tests pass random_state=0 so they cannot flake on an unlucky initialization. Everything else is yours as submitted.

**Comment by wshobson** [comment]
Merged with the two fixups described above. Thank you; #247 closes with it.
